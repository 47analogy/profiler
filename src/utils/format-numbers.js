/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow

/**
 * Format a positive float into a string.
 *
 * Try to format the value to 2 significant digits as much as possible but
 * without using scientific notation.  The number of decimal places depends
 * on the value: the closer to zero the value is, the more decimal places
 * are used in the resulting string.
 *
 * For example:
 *
 * formatNumber(123,   ) = "123"
 * formatNumber(12.3,  ) =  "12"
 * formatNumber(1.23,  ) =   "1.2"
 * formatNumber(0.01234) =   "0.012"
 */
export function formatNumber(value: number): string {
  let result: string;
  if (value >= 10) {
    result = value.toFixed(0);
  } else if (value >= 1) {
    result = value.toFixed(1);
  } else if (value >= 0.1) {
    result = value.toFixed(2);
  } else {
    result = value.toFixed(3);
  }

  return result;
}

export function formatPercent(value: number): string {
  return formatNumber(value * 100) + '%';
}
