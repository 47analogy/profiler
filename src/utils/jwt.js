/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
// @flow

// This file provides simple utils to deal with JWT tokens.

export function extractAndDecodePayload(jwtToken: string): any {
  if (!isValidJwtToken(jwtToken)) {
    console.error('The token is an invalid JWT token.');
    return null;
  }

  try {
    const payload = jwtToken.split('.')[1];
    const decodedPayload = decodeJwtBase64Url(payload);
    const jsonPayload = JSON.parse(decodedPayload);

    return jsonPayload;
  } catch (e) {
    console.error(
      `We got an unexpected error when trying to decode the JWT token '${jwtToken}':`,
      e
    );
    return null;
  }
}

// This uses the base64url characters, that is base64 characters where + is
// replaced by -, and / is replaced by _. Moreover the padding character isn't
// used with JWT.
const JWT_TOKEN_RE = /^(?:[a-zA-Z0-9_-])+\.(?:[a-zA-Z0-9_-])+\.(?:[a-zA-Z0-9_-])+$/;
export function isValidJwtToken(jwtToken: string): boolean {
  return JWT_TOKEN_RE.test(jwtToken);
}

export function decodeJwtBase64Url(base64UrlEncodedValue: string): string {
  // In the base64url variant used in JWT, the padding "=" character is removed.
  // But atob doesn't mind, so we don't need to recover the missing padding like
  // most implementations do.

  // We do need to convert the string to a "normal" base64 encoding though.
  const base64EncodedValue = base64UrlEncodedValue
    .replace('-', '+')
    .replace('_', '/');

  return atob(base64EncodedValue);
}
