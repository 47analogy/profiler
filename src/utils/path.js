/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow

import type { CallNodePath } from '../types/profile-derived';

export function arePathsEqual(a: CallNodePath, b: CallNodePath): boolean {
  if (a === b) {
    return true;
  }

  if (a.length !== b.length) {
    return false;
  }

  // Iterating from the end because this will likely fail faster
  for (let i = a.length - 1; i >= 0; i--) {
    if (a[i] !== b[i]) {
      return false;
    }
  }

  return true;
}

// We take the easy path by converting the path to a string that will be unique.
// In the future and if necessary we might want to have an integer-based hash so
// that it's faster. In that case we might want to use only the n last values
// that should be more discriminate and keep it fast.
export function hashPath(a: CallNodePath): string {
  return a.join('-');
}

export class PathSet implements Iterable<CallNodePath> {
  _table: Map<string, CallNodePath>;

  constructor(iterable?: Iterable<CallNodePath>) {
    if (iterable instanceof PathSet) {
      // This shortcut avoids the call to `hashPath` by taking advantage of
      // knowing some inner workings.
      this._table = new Map(iterable._table);
      return;
    }

    this._table = new Map();

    if (!iterable) {
      return;
    }

    for (const path of iterable) {
      this._table.set(hashPath(path), path);
    }
  }

  add(path: CallNodePath) {
    this._table.set(hashPath(path), path);
    return this;
  }

  clear() {
    this._table.clear();
  }

  delete(path: CallNodePath) {
    return this._table.delete(hashPath(path));
  }

  *values(): Iterator<CallNodePath> {
    yield* this._table.values();
  }

  *entries(): Iterator<[CallNodePath, CallNodePath]> {
    for (const entry of this) {
      yield [entry, entry];
    }
  }

  forEach(func: (CallNodePath, CallNodePath, PathSet) => void, thisArg?: any) {
    for (const entry of this) {
      func.call(thisArg, entry, entry, this);
    }
  }

  has(path: CallNodePath): boolean {
    return this._table.has(hashPath(path));
  }

  get size(): number {
    return this._table.size;
  }

  // https://github.com/facebook/flow/issues/3258
  // $FlowFixMe
  *[Symbol.iterator]() {
    yield* this._table.values();
  }

  // This is a hack for Flow's Iterable support,
  // see https://stackoverflow.com/questions/48491307/iterable-class-in-flow
  /*::
  @@iterator(): * {
    // $FlowFixMe
    return this[Symbol.iterator]()
  }
  */
}
