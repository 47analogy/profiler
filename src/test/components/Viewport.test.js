/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import * as React from 'react';
import { Provider } from 'react-redux';
import { mount } from 'enzyme';

import { withChartViewport } from '../../components/shared/chart/Viewport';
import {
  getCommittedRange,
  getPreviewSelection,
} from '../../reducers/profile-view';
import explicitConnect from '../../utils/connect';
import mockCanvasContext from '../fixtures/mocks/canvas-context';
import mockRaf from '../fixtures/mocks/request-animation-frame';
import { storeWithProfile } from '../fixtures/stores';
import { getBoundingBox, getMouseEvent } from '../fixtures/utils';
import { getProfileFromTextSamples } from '../fixtures/profiles/make-profile';

// Define some magic sizing values to assert various properties of the viewport.
const COMPONENT_WIDTH = 300;
const COMPONENT_HEIGHT = 100;
const LEFT = 5;
const TOP = 7;
const MAXIMUM_ZOOM = 0.1;
const MAX_VIEWPORT_HEIGHT = COMPONENT_HEIGHT * 3;
const SMALL_MAX_VIEWPORT_HEIGHT = COMPONENT_HEIGHT * 0.2;

describe('Viewport', function() {
  it('matches the component snapshot', () => {
    const { view } = setup();
    expect(view).toMatchSnapshot();
    // Trigger any unmounting behavior handlers, just make sure it doesn't
    // throw any errors.
    view.unmount();
  });

  it('renders the wrapped chart component', () => {
    const { view } = setup();
    expect(view.find('#dummy-chart').length).toBe(1);
  });

  it('provides a viewport', () => {
    const { getChartProps } = setup();
    const { viewport } = getChartProps();
    expect(viewport).toEqual({
      containerWidth: COMPONENT_WIDTH,
      containerHeight: COMPONENT_HEIGHT,
      viewportLeft: 0,
      viewportRight: 1,
      viewportTop: 0,
      viewportBottom: COMPONENT_HEIGHT,
      isDragging: false,
      isSizeSet: true,
      // Just pass the function through.
      moveViewport: viewport.moveViewport,
    });
  });

  describe('scrolling hint', function() {
    it('can show a shift scrolling hint', function() {
      jest.useFakeTimers();
      const { view, scroll } = setup();
      const isTimerVisible = () =>
        !view.find('.chartViewportShiftScroll').hasClass('hidden');

      // No hint is shown.
      expect(isTimerVisible()).toBe(false);

      // Scroll a bit to show the menu
      scroll({ deltaY: 10 });
      expect(isTimerVisible()).toBe(true);

      // Run the setTimeout, the menu should disappear.
      jest.runAllTimers();
      view.update();
      expect(isTimerVisible()).toBe(false);
    });

    it('will not show a shift scrolling hint after zooming once', function() {
      const { view, scroll } = setup();
      const isTimerVisible = () =>
        !view.find('.chartViewportShiftScroll').hasClass('hidden');

      // No hint is shown at the beginning.
      expect(isTimerVisible()).toBe(false);

      // Zoom in a bit.
      scroll({ deltaY: 10, shiftKey: true });
      expect(isTimerVisible()).toBe(false);

      // Now scroll, no hint should show.
      scroll({ deltaY: 10 });
      expect(isTimerVisible()).toBe(false);
    });
  });

  describe('scrolling up and down', function() {
    it('scrolls the viewport down using the mousewheel', () => {
      const { view, getChartProps } = setup();
      expect(getChartProps().viewport).toMatchObject({
        viewportLeft: 0,
        viewportRight: 1,
        viewportTop: 0,
        viewportBottom: COMPONENT_HEIGHT,
      });
      const deltaY = 10;
      view.find('.chartViewport').simulate('wheel', getMouseEvent({ deltaY }));
      expect(getChartProps().viewport).toMatchObject({
        viewportLeft: 0,
        viewportRight: 1,
        viewportTop: deltaY,
        viewportBottom: COMPONENT_HEIGHT + deltaY,
      });
    });

    it('cannot scroll past the bottom of the viewport', () => {
      const { view, getChartProps } = setup();
      const deltaY = 1000;
      view.find('.chartViewport').simulate('wheel', getMouseEvent({ deltaY }));
      view.update();
      expect(getChartProps().viewport).toMatchObject({
        viewportLeft: 0,
        viewportRight: 1,
        viewportTop: MAX_VIEWPORT_HEIGHT - COMPONENT_HEIGHT,
        viewportBottom: MAX_VIEWPORT_HEIGHT,
      });
    });

    it('cannot scroll past the top of the viewport', () => {
      const { view, getChartProps } = setup();
      const deltaY = -100;
      view.find('.chartViewport').simulate('wheel', getMouseEvent({ deltaY }));
      view.update();
      expect(getChartProps().viewport).toMatchObject({
        viewportLeft: 0,
        viewportRight: 1,
        viewportTop: 0,
        viewportBottom: COMPONENT_HEIGHT,
      });
    });
  });

  describe('mousewheel zooming using shift + mousewheel', function() {
    // These series of tests are very particular to margin differences, so check
    // different margin combinations separately.
    const marginTests = [
      { marginLeft: 0, marginRight: 0, description: 'no margins' },
      { marginLeft: 33, marginRight: 0, description: 'with a left margin' },
      { marginLeft: 0, marginRight: 33, description: 'with a right margin' },
      { marginLeft: 33, marginRight: 55, description: 'with both margins' },
    ];

    // Run through the tests.
    for (const { marginLeft, marginRight, description } of marginTests) {
      describe(description, function() {
        // Take into account the margins and the component width to find the inner
        // component width, which is the active viewport area. All mouse events should
        // be relative to this space.
        const innerComponentWidth = COMPONENT_WIDTH - marginLeft - marginRight;

        // |        Viewport        |
        // |-----------*------------|
        //             ^
        //             Mouse position
        it('zooms the preview selection equally when the mouse is centered', () => {
          const { scrollAndGetViewport } = setup({
            marginLeft,
            marginRight,
          });
          const viewport = scrollAndGetViewport({
            deltaY: -100,
            shiftKey: true,
            clientX: LEFT + innerComponentWidth * 0.5 + marginLeft,
          });

          // Assert that this zooms in equally.
          expect(viewport.viewportLeft).toBeGreaterThan(0);
          expect(viewport.viewportRight).toBeLessThan(1);
          expect(1 - viewport.viewportRight).toBeCloseTo(
            viewport.viewportLeft,
            7
          );

          // Only do an additional viewport top/bottom check here.
          expect(viewport).toMatchObject({
            viewportTop: 0,
            viewportBottom: COMPONENT_HEIGHT + 0,
          });
        });

        // |        Viewport        |
        // |------------------*-----|
        //                    ^
        //                    Mouse position
        it('zooms the preview selection in a direction when the mouse is to the right', () => {
          const { scrollAndGetViewport } = setup({
            marginLeft,
            marginRight,
          });
          const viewport = scrollAndGetViewport({
            deltaY: -10,
            shiftKey: true,
            clientX: LEFT + innerComponentWidth * 0.75 + marginLeft,
          });

          // Assert that the left hand side zooms in more.
          const changeInLeft = viewport.viewportLeft;
          const changeInRight = 1 - viewport.viewportRight;
          expect(viewport.viewportLeft).toBeGreaterThan(0);
          expect(viewport.viewportRight).toBeLessThan(1);
          expect(changeInLeft).toBeGreaterThan(changeInRight);
        });

        // |        Viewport        |
        // |----*-------------------|
        //      ^
        //      Mouse position
        it('zooms the preview selection in a direction when the mouse is to the left', () => {
          const { scrollAndGetViewport } = setup({
            marginLeft,
            marginRight,
          });
          const viewport = scrollAndGetViewport({
            deltaY: -10,
            shiftKey: true,
            clientX: LEFT + innerComponentWidth * 0.25 + marginLeft,
          });

          // Assert that the left hand side zooms in more.
          const changeInLeft = viewport.viewportLeft;
          const changeInRight = 1 - viewport.viewportRight;
          expect(viewport.viewportLeft).toBeGreaterThan(0);
          expect(viewport.viewportRight).toBeLessThan(1);
          expect(changeInLeft).toBeLessThan(changeInRight);
        });

        // |        Viewport        |
        // |*-----------------------|
        //  ^
        //  Mouse position
        it('does not scroll the viewport left when the mouse is centered on the left.', () => {
          const { scrollAndGetViewport } = setup({
            marginLeft,
            marginRight,
          });
          const viewport = scrollAndGetViewport({
            deltaY: -10,
            shiftKey: true,
            clientX: LEFT + marginLeft,
          });
          expect(viewport.viewportLeft).toBe(0);
          expect(viewport.viewportRight).toBeLessThan(1);
        });

        // |        Viewport        |
        // |-----------------------*|
        //                         ^
        //                         Mouse position
        it('does not scroll the viewport right when the mouse is centered on the right.', () => {
          const { scrollAndGetViewport } = setup({
            marginLeft,
            marginRight,
          });
          const viewport = scrollAndGetViewport({
            deltaY: -10,
            shiftKey: true,
            clientX: LEFT + innerComponentWidth + marginLeft,
          });
          expect(viewport.viewportLeft).toBeGreaterThan(0);
          expect(viewport.viewportRight).toBe(1);
        });

        it('cannot zoom out beyond the bounds', () => {
          const { scrollAndGetViewport } = setup({
            marginLeft,
            marginRight,
          });
          const viewport = scrollAndGetViewport({
            deltaY: 100,
            shiftKey: true,
            clientX: LEFT + innerComponentWidth * 0.5 + marginLeft,
          });
          expect(viewport).toMatchObject({
            viewportLeft: 0,
            viewportRight: 1,
          });
        });
      });
    }
  });

  describe('dragging around', function() {
    const middleX = COMPONENT_WIDTH * 0.5;
    const middleY = COMPONENT_HEIGHT * 0.5;

    it('can click and drag down', function() {
      const { getChartProps, clickAndDrag } = setup();

      expect(getChartProps().viewport).toMatchObject({
        viewportLeft: 0,
        viewportRight: 1,
        viewportTop: 0,
        viewportBottom: COMPONENT_HEIGHT,
      });
      clickAndDrag(middleX, middleY, middleX, middleY - 50);
      expect(getChartProps().viewport).toMatchObject({
        viewportLeft: 0,
        viewportRight: 1,
        viewportTop: 50,
        viewportBottom: COMPONENT_HEIGHT + 50,
      });
    });

    it('can click and drag up', function() {
      const { getChartProps, clickAndDrag } = setup();
      const middleX = COMPONENT_WIDTH * 0.5;
      const middleY = COMPONENT_HEIGHT * 0.5;

      clickAndDrag(middleX, middleY, middleX, middleY - 100);
      expect(getChartProps().viewport).toMatchObject({
        viewportLeft: 0,
        viewportRight: 1,
        viewportTop: 100,
        viewportBottom: COMPONENT_HEIGHT + 100,
      });
      clickAndDrag(middleX, middleY, middleX, middleY + 50);
      expect(getChartProps().viewport).toMatchObject({
        viewportLeft: 0,
        viewportRight: 1,
        viewportTop: 50,
        viewportBottom: COMPONENT_HEIGHT + 50,
      });
    });

    it('will not drag beyond the top of the window', function() {
      const { getChartProps, clickAndDrag } = setup();
      const middleX = COMPONENT_WIDTH * 0.5;
      const middleY = COMPONENT_HEIGHT * 0.5;

      clickAndDrag(middleX, middleY, middleX, middleY + 50);
      expect(getChartProps().viewport).toMatchObject({
        viewportLeft: 0,
        viewportRight: 1,
        viewportTop: 0,
        viewportBottom: COMPONENT_HEIGHT,
      });
    });

    it('will not drag beyond the bottom of the window', function() {
      const { getChartProps, clickAndDrag } = setup();
      const middleX = COMPONENT_WIDTH * 0.5;
      const middleY = COMPONENT_HEIGHT * 0.5;

      clickAndDrag(
        middleX,
        middleY,
        middleX,
        middleY - MAX_VIEWPORT_HEIGHT * 2
      );
      expect(getChartProps().viewport).toMatchObject({
        viewportLeft: 0,
        viewportRight: 1,
        viewportTop: MAX_VIEWPORT_HEIGHT - COMPONENT_HEIGHT,
        viewportBottom: MAX_VIEWPORT_HEIGHT,
      });
    });

    it('will anchor small content at the top', function() {
      const { getChartProps, clickAndDrag } = setup({
        maxViewportHeight: SMALL_MAX_VIEWPORT_HEIGHT,
        startsAtBottom: false,
      });
      const anchoredViewport = {
        viewportLeft: 0,
        viewportRight: 1,
        viewportTop: 0,
        viewportBottom: COMPONENT_HEIGHT,
      };
      expect(getChartProps().viewport).toMatchObject(anchoredViewport);

      // It will not move the viewport.
      clickAndDrag(middleX, middleY, middleX, middleY - 10);
      expect(getChartProps().viewport).toMatchObject(anchoredViewport);
      clickAndDrag(middleX, middleY, middleX, middleY + 10);
      expect(getChartProps().viewport).toMatchObject(anchoredViewport);
    });

    it('will anchor small content at the bottom', function() {
      const { getChartProps, clickAndDrag } = setup({
        maxViewportHeight: SMALL_MAX_VIEWPORT_HEIGHT,
        startsAtBottom: true,
      });
      const anchoredViewport = {
        viewportLeft: 0,
        viewportRight: 1,
        viewportTop: SMALL_MAX_VIEWPORT_HEIGHT - COMPONENT_HEIGHT,
        viewportBottom: SMALL_MAX_VIEWPORT_HEIGHT,
      };
      expect(getChartProps().viewport).toMatchObject(anchoredViewport);

      // It will not move the viewport.
      clickAndDrag(middleX, middleY, middleX, middleY - 10);
      expect(getChartProps().viewport).toMatchObject(anchoredViewport);
      clickAndDrag(middleX, middleY, middleX, middleY + 10);
      expect(getChartProps().viewport).toMatchObject(anchoredViewport);
    });

    it('can click and drag left/right', function() {
      const { scrollAndGetViewport, getChartProps, clickAndDrag } = setup();

      // Assert the initial values.
      const { viewportLeft, viewportRight } = scrollAndGetViewport({
        // Zoom in some large arbitrary amount:
        deltaY: -5000,
        shiftKey: true,
        clientX: LEFT + COMPONENT_WIDTH * 0.5,
      });

      // These values are arbitrary, but show that the viewport was zoomed in.
      expect(viewportLeft).toBeGreaterThan(0.3);
      expect(viewportRight).toBeLessThan(0.7);

      // Perform the dragging action.
      clickAndDrag(middleX, middleY, middleX + 500, middleY);
      expect(getChartProps().viewport.viewportLeft).toBeLessThan(viewportLeft);
      expect(getChartProps().viewport.viewportRight).toBeLessThan(
        viewportRight
      );

      // Drag back the other way.
      clickAndDrag(middleX, middleY, middleX - 1000, middleY);
      expect(getChartProps().viewport.viewportLeft).toBeGreaterThan(
        viewportLeft
      );
      expect(getChartProps().viewport.viewportRight).toBeGreaterThan(
        viewportRight
      );
    });

    it('will not scroll off to the left of the viewport bounds', function() {
      const { scrollAndGetViewport, getChartProps, clickAndDrag } = setup();
      scrollAndGetViewport({
        // Zoom in some large arbitrary amount:
        deltaY: -5000,
        shiftKey: true,
        clientX: LEFT + COMPONENT_WIDTH * 0.5,
      });

      // Perform the dragging action some arbitrarily large distance..
      clickAndDrag(middleX, middleY, middleX + 10000, middleY);
      expect(getChartProps().viewport.viewportLeft).toBe(0);
      expect(getChartProps().viewport.viewportRight).toBeLessThan(0.4);
    });

    it('will not scroll off to the right of the viewport bounds', function() {
      const { scrollAndGetViewport, getChartProps, clickAndDrag } = setup();

      scrollAndGetViewport({
        // Zoom in some large arbitrary amount:
        deltaY: -5000,
        shiftKey: true,
        clientX: LEFT + COMPONENT_WIDTH * 0.5,
      });

      // Perform the dragging action some arbitrarily large distance.
      clickAndDrag(middleX, middleY, middleX - 10000, middleY);
      expect(getChartProps().viewport.viewportLeft).toBeGreaterThan(0.6);
      expect(getChartProps().viewport.viewportRight).toBe(1);
    });
  });
});

function setup(profileOverrides: Object = {}) {
  const flushRafCalls = mockRaf();
  const ctx = mockCanvasContext();

  jest
    .spyOn(HTMLCanvasElement.prototype, 'getContext')
    .mockImplementation(() => ctx);

  jest
    .spyOn(HTMLElement.prototype, 'getBoundingClientRect')
    .mockImplementation(() => {
      const rect = getBoundingBox(COMPONENT_WIDTH, COMPONENT_HEIGHT);
      // Add some arbitrary X offset.
      rect.left += LEFT;
      rect.right += LEFT;
      rect.x += LEFT;
      rect.y += TOP;
      rect.top += TOP;
      rect.bottom += TOP;
      return rect;
    });

  // Hook up a dummy chart with a viewport.
  const DummyChart = () => <div id="dummy-chart" />;
  const ChartWithViewport = withChartViewport(DummyChart);

  // This component assumes that it is inside of a connected component, so go
  // ahead and mock that out here.
  const ConnectedChartWithViewport = explicitConnect({
    mapStateToProps: state => ({
      timeRange: getCommittedRange(state),
      previewSelection: getPreviewSelection(state),
    }),
    mapDispatchToProps: {},
    // eslint-disable-next-line react/display-name
    component: (props: Object) => (
      <ChartWithViewport
        viewportProps={{
          previewSelection: props.previewSelection,
          timeRange: props.timeRange,
          maxViewportHeight: props.maxViewportHeight || MAX_VIEWPORT_HEIGHT,
          startsAtBottom: props.startsAtBottom,
          viewportNeedsUpdate: () => false,
          marginLeft: 0,
          marginRight: 0,
          maximumZoom: MAXIMUM_ZOOM,
          ...profileOverrides,
        }}
        chartProps={
          {
            // None used in dummy component.
          }
        }
      />
    ),
  });

  const view = mount(
    <Provider store={storeWithProfile(getProfileFromTextSamples('A').profile)}>
      <ConnectedChartWithViewport />
    </Provider>
  );

  // WithSize uses requestAnimationFrame.
  flushRafCalls();
  view.update();

  // The following functions are helpers for the tests, to provide a nicer functional
  // interface to drive changes to the components.

  function moveMouseAndGetLeft(pageX: number): number {
    view.simulate('mousemove', { pageX });
    view.update();
    return view.find('.timelineTrackScreenshotHover').prop('style').left;
  }

  function getChartProps() {
    return view.find(DummyChart).props();
  }

  function scroll(eventOverrides) {
    view
      .find('.chartViewport')
      .simulate('wheel', getMouseEvent(eventOverrides));
    flushRafCalls();
    view.update();
  }

  function scrollAndGetViewport(eventOverrides) {
    scroll(eventOverrides);
    return getChartProps().viewport;
  }

  // Dispatch events to the window, which isn't available through the enzyme
  // simulate interface. Instead use JSDOM.
  function _dispatchToJsdomWindow(eventType: string, overrides: Object) {
    window.dispatchEvent(
      // Flow doesn't like us adding unknown mouse properties to Event, and also
      // doesn't like us modifying a MouseEvent, so opt out of type checks here.
      Object.assign((new Event(eventType): Object), getMouseEvent(overrides))
    );
  }

  function clickAndDrag(fromX, fromY, toX, toY) {
    const from = { clientX: fromX, clientY: fromY };
    const to = { clientX: toX, clientY: toY };

    view.find('.chartViewport').simulate('mousedown', getMouseEvent(from));
    _dispatchToJsdomWindow('mousemove', from);
    _dispatchToJsdomWindow('mousemove', to);
    _dispatchToJsdomWindow('mouseup', to);
    // Flush any batched updates.
    flushRafCalls();
    view.update();
  }

  return {
    view,
    moveMouseAndGetLeft,
    flushRafCalls,
    getChartProps,
    scrollAndGetViewport,
    scroll,
    clickAndDrag,
  };
}
