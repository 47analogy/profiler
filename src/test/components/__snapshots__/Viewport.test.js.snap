// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Viewport matches the component snapshot 1`] = `
<div
  class="chartViewport"
  tabindex="0"
>
  <div
    id="dummy-chart"
  />
  <div
    class="chartViewportCtrlScroll hidden"
  >
    Zoom Chart:
    <kbd
      class="chartViewportCtrlScrollKbd"
    >
      Ctrl
    </kbd>
    <kbd
      class="chartViewportCtrlScrollKbd"
    >
      Scroll
    </kbd>
  </div>
</div>
`;
